src/libbson/src/bson/bson-dsl.h

```diff
@@ -0,0 +1,1336 @@
+         sizeof _bpVisitBits_static / sizeof (uint64_t);                       \
+      BSON_MAYBE_UNUSED bool _bpFoundElement = false;                          \
+      _bsonParse_applyOps (__VA_ARGS__);                                       \
+      /* We may have allocated for visit bits */                               \
+      if (_bpVisitBits != _bpVisitBits_static) {                               \
+         bson_free (_bpVisitBits);                                             \
+      }                                                                        \
+      /* Restore the dsl context */                                            \
+      _bsonVisitContextThreadLocalPtr = bsonVisitContext.parent;               \
+   } while (0)
+
+#define _bsonParse_applyOps(...) \
+   _bsonDSL_mapMacro (_bsonParse_applyOp, ~, __VA_ARGS__)
+
+/// Parse one entry referrenced by the context iterator
+#define _bsonParse_applyOp(P, _nil, Counter) \
+   do {                                      \
+      if (!_bvHalt && !bsonParseError) {     \
+         _bsonParseOperation_##P;            \
+      }                                      \
+   } while (0);
+
+#define _bsonParseMarkVisited(Index)                                   \
+   if (1) {                                                            \
+      const int nth_int = Index / 64;                                  \
+      const int nth_bit = Index % 64;                                  \
+      while (nth_int >= _bpNumVisitBitInts) {                          \
+         /* Say that five times, fast: */                              \
+         int new_num_visit_bit_ints = _bpNumVisitBitInts * 2;          \
+         uint64_t *new_visit_bit_ints =                                \
+            bson_malloc0 (sizeof (uint64_t) * new_num_visit_bit_ints); \
+         memcpy (new_visit_bit_ints,                                   \
+                 _bpVisitBits,                                         \
+                 sizeof (uint64_t) * _bpNumVisitBitInts);              \
+         if (_bpVisitBits != _bpVisitBits_static) {                    \
+            bson_free (_bpVisitBits);                                  \
+         }                                                             \
+         _bpVisitBits = new_visit_bit_ints;                            \
+         _bpNumVisitBitInts = new_num_visit_bit_ints;                  \
+      }                                                                \
+                                                                       \
+      _bpVisitBits[nth_int] |= (1 << nth_bit);                         \
+   } else                                                              \
+      ((void) 0)
+
+#define _bsonParseDidVisitNth(Index) \
+   _bsonParseDidVisitNth_1 (Index / 64, Index % 64)
+#define _bsonParseDidVisitNth_1(NthInt, NthBit) \
+   (NthInt < _bpNumVisitBitInts && (_bpVisitBits[NthInt] & (1 << NthBit)))
```

Comment:
```suggestion
#define _bsonParseDidVisitNth(Index) \
   _bsonParseDidVisitNth_1 (Index / 64u, Index % 64u)
#define _bsonParseDidVisitNth_1(NthInt, NthBit) \
   (NthInt < _bpNumVisitBitInts &&              \
    (_bpVisitBits[NthInt] & (UINT64_C (1) << NthBit)))
```
Type consistency with `_bsonVisitContext_t::index` and `_bpVisitbits[NthBit]`.